tests/deepsparse/transformers/pipelines/test_text_generation.py

```diff
@@ -12,262 +12,396 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from typing import List, Optional, Tuple
 
-import numpy as np
-import onnx
-import onnxruntime
-from transformers import AutoModelForCausalLM, AutoTokenizer
+import numpy
 
 import pytest
 from deepsparse import Pipeline
-from deepsparse.transformers.utils.helpers import create_causal_mask
-from deepsparse.utils.onnx import (
-    CACHE_INPUT_PREFIX,
-    overwrite_onnx_model_inputs_for_kv_cache_models,
+from deepsparse.transformers.utils.decoder_kv_cache import DecoderKVCache
+from tests.deepsparse.transformers.pipelines.helpers import (
+    ORTGroundTruthSource,
+    TorchGroundTruthSource,
 )
-from sparsezoo import Model
-
-
-def _initialize_kv_cache_state(model, length=0):
-    # get one of the cache inputs
-    cache_input = next(
-        input
-        for input in model.graph.input
-        if input.name.startswith(CACHE_INPUT_PREFIX)
-    )
-    # read the shape of the cache input
-    batch_size = cache_input.type.tensor_type.shape.dim[0].dim_value
-    num_attention_heads = cache_input.type.tensor_type.shape.dim[1].dim_value
-    hidden_dims = cache_input.type.tensor_type.shape.dim[3].dim_value
-
-    # create a kv cache dictionary
-    kv_cache = {
-        input_.name: np.zeros(
-            (batch_size, num_attention_heads, length, hidden_dims), dtype=np.float32
-        )
-        for input_ in model.graph.input
-        if input_.name.startswith(CACHE_INPUT_PREFIX)
-    }
-
-    return kv_cache
-
-
-START = 0  # global variable for dummy_callback
 
 
 @pytest.mark.parametrize(
-    "internal_kv_cache",
+    "use_deepsparse_cache",
     [True, False],
 )
 @pytest.mark.parametrize(
-    "model_stub, model_name, uses_bos_token",
+    "model_stub, model_name, uses_bos_token, logits_max_diff_kv_cache_has_been_filled",
     [
-        (
-            "zoo:nlg/text_generation/opt-1.3b/pytorch/"
-            "huggingface/opt_pretrain/base-none",
-            "facebook/opt-1.3b",
-            True,
-        ),
         (
             "zoo:nlg/text_generation/codegen_mono-350m/pytorch/"
             "huggingface/bigpython_bigquery_thepile/base-none",
             "salesforce/codegen-350m-mono",
```

Comment:
If we're using codegen would it make sense to add a codegen-specific prompt? It may not matter since we're comparing it to the groundtruth from hugging face but both will essentially produce garbage for the current prompt?

Reply:
Hmm, adds a bit more complexity to the whole harness, but you have a good point there.

Reply:
Acknowledged